src/libsyntax/parse/token.rs

```diff
@@ -476,6 +476,19 @@ impl Token {
         false
     }
 
+    /// Would `maybe_whole_expr` in `parser.rs` return `Ok(..)`?
+    /// That is, is this a pre-parsed expression dropped into the token stream
+    /// (which happens while parsing the result ofmacro expansion)?
+    crate fn is_whole_expr(&self) -> bool {
+        if let Interpolated(ref nt) = self.kind {
+            if let NtExpr(_) | NtLiteral(_) | NtPath(_) | NtBlock(_) = **nt {
```

Comment:
Should this include `NtIdent` and should `parse_prefix_expr` use `is_whole_expr` in the case of `token::Ident(..) if self.token.is_ident_named(sym::not) => {`?

(Might want to look over the usage of `Interpolated` in `token.rs` and `parser.rs` overall.)

Reply:
>Should this include `NtIdent`

`NtIdent` is normalized to `Ident` in `Parser::bump`, so only "parser infrastructure" code need to care about it, but not "parser user" code, so `macro_rules! maybe_whole_expr` don't need it.

However, since you added this method to `Token` (which is not necessarily tied to the `Parser` in theory), adding `NtIdent` here won't hurt.

`parse_prefix_expr` can reuse `is_whole_expr`, yes.